### PR TITLE
Improve printing of strings

### PR DESCRIPTION
**Pull request by som-snytt**
Ports https://github.com/scala/scala/pull/10897

It would be nice if SourceCode avoided intermediate strings (due to highlighting) and used the StringBuilder directly.

### CODE REVIEW

**Comment by som-snytt** [comment]
`cc`

**Comment by Mikhail42** [comment]
It would be nice to add at least a few tests. I'he not found them here (https://github.com/scala/scala3/tree/main/compiler/test/dotty/tools/dotc/printing)

**Comment by som-snytt** [comment]
I would do something about the code duplication before undrafting.

Everything breaks if strings are broken, and on scala 2 it was also a zinc dependency. A useful test would be a scalacheck, as there are many boundary conditions. An existing test broke because it noticed whether Unicode escapes were lowercase. (I did not argue with the test.)

File: compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala [hunk 1]
```diff
@@ -731,7 +729,57 @@ class PlainPrinter(_ctx: Context) extends Printer {
...
+      while i < text.length do
+        escapedChar(b, text.charAt(i))
+        i += 1
+      if quoted then
+        b.append('"')
+      b.toString
+    else if quoted then "\"" + text + "\""
+    else text
+
+  private def escapedChar(b: StringBuilder, c: Char): Unit =
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala]: Can we move the functions to `compiler/src/dotty/tools/dotc/util/Chars.scala` to avoid duplication?

File: compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala [hunk 2]
```diff
@@ -731,7 +729,57 @@ class PlainPrinter(_ctx: Context) extends Printer {
...
+    def quadNibble(b: StringBuilder, x: Int, i: Int): Unit =
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala]: We don't need the parameter `b` here, since it is not changed.

File: compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala (hunk 1 above)
Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala]: I guess we can!

**Comment by som-snytt** [comment]
Inlined `quadNibble` just for the pleasure of watching it unroll. It would be nice to invest in a jmh test to verify that it doesn't have unintended consequences. I misplaced the test for Scala 2 that demonstrated it.

**Comment by som-snytt** [comment]
rebased
